Course Creation and Specification

spencerj411 opened the pull request:
Completed Issue #27 (Course Creation and Specification (5)).

Can access this page at `http://127.0.0.1:8000/manage_courses/`. It displays a form in 3 sections (for this user story, only 'Add Course' has been implemented).

**The acceptance criterion:** 
> If the requisite/incompatible courses aren’t in the system after the user saves the new course, the user is prompted to fill in the course details for the requisite/incompatible course

has been satisfied by listing all of the courses in the database instead of giving the option to the user to type the courses manually.
> Notify the user if they are attempting to add an existing course

was implemented with a basic HttpResponse.

**Small things:**
1. I was not sure how to concatenate 3 strings together hence `{{ course.code }}_{{ course.year }}` appeared 4 times in `managecourses.html` (solved)
2. Creation of courses can be tested by checking `http://127.0.0.1:8000/list/?view=Course`
3. Screenshot of UI  (outdated, check below for a more updated screenshot)
![image](https://user-images.githubusercontent.com/24206502/56049550-15431300-5d8d-11e9-988d-e3588fb17215.png)

spencerj411 committed:
Created managecourses.html file based of sampleform.html, created manage_courses views function and linked io to urls.py

spencerj411 committed:
corrected page-title typo

spencerj411 committed:
corrected Add Course typo

spencerj411 committed:
added requisite and incompatible courses feature

spencerj411 committed:
added extra comments

JackSekoranja reviewed (changes requested):
Form looks really good, but since adding and editing should be the same form we may be able to simplify it using a similar trick to my `list` page, so I think leave the additional forms out of it for now. There were also a few minor changes that I will suggest below.

JackSekoranja commented on cassdegrees/ui/views.py:
```diff
@@ -95,3 +95,66 @@ def sampleform(request):
+    # If POST request, redirect the received information to the backend:
+    if request.method == 'POST':
+        # hard coded url; only temporary
+        model_api_url = 'http://127.0.0.1:8000/api/model/course/'
```
```suggestion
        model_api_url = request.build_absolute_uri('/api/model/course/')
```

This Makes it a relative path, meaning we won't have to change it later (We should probably do this in `sampleform` above too)

JackSekoranja commented on cassdegrees/templates/managecourses.html:
```diff
@@ -0,0 +1,115 @@
+            <p>
+                <label for="text">Year:</label>
+                <!-- static min max values are only temporary -->
+                <input class="text" id="number" type="number" name="year" min="1990" max="2100" aria-required="true" required>
```
```suggestion
                <input class="text tfull" id="number" type="number" name="year" min="1990" aria-required="true" required>
```

Changes will make the form stretch across to the end of the page. Also remove max year, it isn't needed.

spencerj411 replied:
my thought process was so it leaves the user less of a chance to create a mistake when creating a course, hence adding max for both year. My original intention was to make it dynamic but will do that at a later date once the main work has been approved and if everyone is ok with this.

JackSekoranja replied:
I still don't think we should restrict the software like that though. Maybe we could ask the user for confirmation if the entered year is too distant from the current year (Implementation may be too complicated for the resulting gain though)

JackSekoranja commented on cassdegrees/templates/managecourses.html:
```diff
@@ -0,0 +1,115 @@
+
+            <p>
+                <label for="id">Course Name:</label>
+                <input class="text" id="text" type="text" name="name" aria-required="true" required>
```
```suggestion
                <input class="text tfull" id="text" type="text" name="name" aria-required="true" required>
```

Makes the text box stretch to the end of the page, giving the user more space to type

spencerj411 replied:
good idea

JackSekoranja commented on cassdegrees/templates/managecourses.html:
```diff
@@ -0,0 +1,115 @@
+
+            <p>
+                <label for="id">Course Code:</label>
+                <input class="text" id="text" type="text" name="code" minlength="8" maxlength="8" size="8" aria-required="true" required>
```
```suggestion
                <input class="text tfull" id="text" type="text" name="code" minlength="8" maxlength="8" size="8" aria-required="true" required>
```

Same as above (for consistency)

spencerj411 replied:
not particularly fond of having all the textboxes stretch to the other side of the page, especially when some input don't need to be that long. Having said that, after visiting [http://webpublishing.anu.edu.au/web-style-guide/forms.php](url) it seems like ANU likes it so I'm ok with this.

JackSekoranja replied:
Alternatively maybe you could reposition year and units to be at the bottom and leave those smaller, while making the other text fields/select fields larger. Might need further discussion with team

JackSekoranja commented on cassdegrees/templates/managecourses.html:
```diff
@@ -0,0 +1,115 @@
+
+            <p>
+                <label for="text">Semester(s):</label>
+                <select name="semester[]" size="2" multiple>
```
I can't work out how to make this stretch to the end of the screen, but it would be good to work out

spencerj411 replied:
will do

JackSekoranja replied:
So it looks like the ANU `tfull` CSS just makes the width of text boxes 70%. Maybe a solution is to create a custom CSS field doing the same thing for select boxes

JackSekoranja commented on cassdegrees/templates/managecourses.html:
```diff
@@ -0,0 +1,115 @@
+             <p>
+                <label for="text">Units:</label>
+                 <!-- max value is chosen arbitrarily -->
+                <input class="text" id="number" type="number" name="units" min="6" max="48" step="6" aria-required="true" required>
```
```suggestion
                <input class="text tfull" id="number" type="number" name="units" min="0" step="6" aria-required="true" required>
```

Courses can be worth 0 units, so this should be reflected here.
Stretch to end of form and removed max units for a course (Realistically it should only be 12-24 max but why restrict them?)

spencerj411 replied:
my thought process was so it leaves the user less of a chance to create a mistake when creating a course, hence adding max for both year and units.

JackSekoranja commented on cassdegrees/templates/managecourses.html:
```diff
@@ -0,0 +1,115 @@
+                <label for="text">Requisite Course(s):</label>
+                <select name="semester[]" size="5" multiple>
+                    {% for course in courses %}
+                        <option value="{{ course.code }}_{{ course.year }}">{{ course.code }}_{{ course.year }}</option>
```
```suggestion
                        <option value="{{ course.code }}">{{ course.code }}</option>
```

Course year is not relevant to requisites, so I have removed them (Need to ensure this is also done when passing the API to prevent repeat values). This also raises another important issue, which is that the requisites is not currently represented in the database. This may be a problem for @aidengeorge97 to solve

JackSekoranja commented on cassdegrees/templates/managecourses.html:
```diff
@@ -0,0 +1,115 @@
+                <label for="text">Incompatible Course(s):</label>
+                <select name="semester[]" size="5" multiple>
+                    {% for course in courses %}
+                        <option value="{{ course.code }}_{{ course.year }}">{{ course.code }}_{{ course.year }}</option>
```
```suggestion
                        <option value="{{ course.code }}">{{ course.code }}</option>
```

As above

JackSekoranja commented on cassdegrees/templates/managecourses.html:
```diff
@@ -0,0 +1,115 @@
+        </p>
+    </form>
+    <br />
+    <!-- not yet implemented -->
```
Might be a good idea to remove everything after this line until the `{% endblock %}` line. `Delete` could have it's own form, but the `edit` form would be the exact same as the one above, only with pre-filled values and with the `value=patch` item. Ask me if unsure about how to approach for future stories

JackSekoranja commented on cassdegrees/templates/managecourses.html:
```diff
@@ -0,0 +1,115 @@
+        <fieldset>
+            <legend>Add Course</legend>
+
+            <!-- As HTML forms only support get or post methods,
```
General note about comments in this page. When you comment with `<!-- ... -->` it appears when you view the page source, but when you use Django Template Comments (`{# ... #}`) they get deleted when the page is hosted.

JackSekoranja commented on cassdegrees/ui/views.py:
```diff
@@ -95,3 +95,66 @@ def sampleform(request):
+
+# inspired by the samepleform function creatd by Daniel Jang
+def manage_courses(request):
+    courses = requests.get(request.build_absolute_uri('/api/model/course/?format=json')).json()
```
```suggestion
    courses = requests.get(request.build_absolute_uri('/api/model/course/?format=json')).json()
    courses = [{'code': course} for course in  set([x['code'] for x in courses])]
```

Convert courses to a unique set of only course codes. Please note that this is NOT a good solution and is intended to be a single line fix. Best way might be to try do a SQL Query for unique course codes, consult @Daniel-Jang for this

spencerj411 replied:
will do

JackSekoranja reviewed (changes requested):
These are my suggested changes

JackSekoranja committed:
made model_api_url a relative path

Co-Authored-By: spencerj411 <spencer.j411@gmail.com>

spencerj411 commented:
UI Screenshot after Jack's suggestions
![image](https://user-images.githubusercontent.com/24206502/56073198-831a2980-5de3-11e9-88f2-0463c270bae2.png)

JackSekoranja committed:
added tfull class to course name input box

Co-Authored-By: spencerj411 <spencer.j411@gmail.com>

JackSekoranja committed:
added tfull class to course code input box

Co-Authored-By: spencerj411 <spencer.j411@gmail.com>

JackSekoranja committed:
Removed course.year, Requisite Course(s) only displays course.code

Co-Authored-By: spencerj411 <spencer.j411@gmail.com>

JackSekoranja committed:
Removed course.year, Incompatible Course(s) only displays course.code

Co-Authored-By: spencerj411 <spencer.j411@gmail.com>

JackSekoranja committed:
Updated min value of units from 6 to 0

Co-Authored-By: spencerj411 <spencer.j411@gmail.com>

JackSekoranja committed:
Made manage_courses function pass only the course codes to managecourses.html; (temporary solution)

Co-Authored-By: spencerj411 <spencer.j411@gmail.com>

pep8speaks commented:
Hello @spencerj411! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`cassdegrees/ui/views.py`](https://github.com/cass-degrees/CASS-Degrees-Code/blob/21fc6bfdd887d3c555ddc0cff6f69043f2cc2425/cassdegrees/ui/views.py):

> [Line 103:46](https://github.com/cass-degrees/CASS-Degrees-Code/blob/21fc6bfdd887d3c555ddc0cff6f69043f2cc2425/cassdegrees/ui/views.py#L103): [E271](https://duckduckgo.com/?q=pep8%20E271) multiple spaces after keyword

spencerj411 committed:
added css code to fix width of Semester(s), Requisite Course(s) and Incompatible Course(s)

spencerj411 committed:
fixed width of year input box to match the other input boxes

spencerj411 commented:
UI after implementing Jack's suggestions
![image](https://user-images.githubusercontent.com/24206502/56073877-71d61a80-5ded-11e9-89ab-0f38632f8d74.png)

spencerj411 committed:
changed select.tfull width so it is aligned with the rest of the input boxes

spencerj411 committed:
added 'required' ANU styling to required fields

JackSekoranja reviewed (approved):
Changes look good. There is an issue where the CSS for `<select>` objects look misaligned but not sure if we can do much about it.

spencerj411 committed:
removed colons after ever label

JackSekoranja committed:
Adds asterisk for required fields and formats forms correctly

• Add default ANU asterisk for required fields
• set up the forms for each operation (Add, Edit, Delete) using the standard form (Comment out the TODO to test)

j-selby commented on cassdegrees/static/css/style.css:
```diff
@@ -27,3 +27,11 @@
+}
+
+select.text{
+    border-radius: 3px;
```
Not specific to this PR, but:

http://webpublishing.anu.edu.au/web-style-guide/style-sampler.php
`Please avoid using inline styles or creating custom style sheets`

We should be justifying each use of rules in this file.

spencerj411 replied:
Not sure if it was mentioned above but it didn't seem like there was a `<select> `version of `tfull` in their css file hence the creation of new rules.

j-selby commented on cassdegrees/ui/views.py:
```diff
@@ -95,3 +95,77 @@ def sampleform(request):
+
+    action = 'Add'; method = 'post' # TODO Remove once other methods are implemented
+
+    courses = requests.get(request.build_absolute_uri('/api/model/course/?format=json')).json()
```
Wait, is this request made even for POST requests, which don't appear to use 'courses' at all? This doesn't seem right.

spencerj411 replied:
@JackSekoranja

j-selby replied:
No longer relevant - main template is now always rendered.

spencerj411 committed:
implemented proper error and success message when form gets submitted. Corrected select name attribute typos.

spencerj411 committed:
corrected semester[] to semesters[], related to last commit

spencerj411 commented:
also added a proper implementation of the error/success message when submitting a form (based off #59) using ANU's style guide.

![image](https://user-images.githubusercontent.com/24206502/56090062-01f48c80-5ee0-11e9-811b-7e8298c096bc.png)

spencerj411 committed:
added success/error message for edit and modify feature of form

j-selby commented on cassdegrees/ui/views.py:
```diff
@@ -95,3 +95,86 @@ def sampleform(request):
+                render_properties['is_error'] = True
+                render_properties['msg'] = \
+                    "Failed to delete course. An unknown error has occurred. Please try again."
+    return render(request, 'managecourses.html',
```
Fix up spacing here?

j-selby commented on cassdegrees/ui/views.py:
```diff
@@ -95,3 +95,86 @@ def sampleform(request):
+
+        # This method of transferring data to the API was inspired by:
+        # https://stackoverflow.com/questions/11663945/calling-a-rest-api-from-django-view
+        if actual_request == "post":
```
We shouldn't be using this for these forms - we only need POST and GET. Get rid of this if statement entirely IMO.

Daniel-Jang replied:
actual_request is now deleted.

Daniel-Jang committed:
Summary: actual_method is now fully abolished, and instead are appending the type of method to the link by adding '?action=Add' for example rather than creating a hidden form to specify type of desired HTML request.
Rewritten some of the manage_courses function in ui/views.py to use the 'action' attribute instead.

Daniel-Jang committed:
Removed unnecessary lines

Daniel-Jang reviewed (changes requested):
This form, as is, can't handle course prereqs properly as they are currently done (e.g. COMPXXXX and COMPXXXX or COMPXXXX).

Also, this needs to be tabbed with Edit, Add and Delete tabs with this spec.:

"When you click on manage courses, there are 3 tabs: Edit, Add and Delete. Screen defaults to edit courses tab, which would have the table of courses with a link that brings out a form that you can change and click submit.
Add tab would just be the Spencer's page in it's current form (look at most recent pic)
Delete would be like the edit page but instead the link submits a delete request to the API."

I have already had conversations with Jack and Spencer and they believe it is a good idea. Spencer has already started implementing this and I will be trying to help him so I can work on my user story: https://github.com/cass-degrees/CASS-Degrees-Code/issues/31

EDIT: Had a chat with Jack and we will have to have a discussion later with Aiden on how to store the requisites properly. Currently, there is no method of storing it properly. Hence, I will approve Spencer's ticket as there is nothing he can do in this moment.

spencerj411 committed:
added tabs in managecourses.html, ready for 'edit' and 'delete' user stories

spencerj411 committed:
removed outdated comments

spencerj411 committed:
fixed up some spacing
